includes/api/class-wc-rest-products-controller.php

```diff
@@ -1422,239 +1460,146 @@ protected function save_variations_data( $product, $request, $single_variation =
-						$skip_image = false;
-
-						if ( 0 === $attachment_id && isset( $image['src'] ) ) {
-							$upload = wc_rest_upload_image_from_url( wc_clean( $image['src'] ) );
-
-							if ( is_wp_error( $upload ) ) {
-								if ( ! apply_filters( 'woocommerce_rest_suppress_variation_image_upload_error', false, $upload, $product ) ) {
-									throw new WC_REST_Exception( 'woocommerce_product_image_upload_error', $upload->get_error_message(), 400 );
-								} else {
-									$skip_image = true;
-								}
-							} else {
-								$attachment_id = wc_rest_set_uploaded_image_as_attachment( $upload, $product->get_id() );
-							}
-						}
-
-						if ( ! $skip_image ) {
-							// Set the image alt if present.
-							if ( ! empty( $image['alt'] ) ) {
-								update_post_meta( $attachment_id, '_wp_attachment_image_alt', wc_clean( $image['alt'] ) );
-							}
-
-							// Set the image name if present.
-							if ( ! empty( $image['name'] ) ) {
-								wp_update_post( array( 'ID' => $attachment_id, 'post_title' => $image['name'] ) );
-							}
-
-							update_post_meta( $variation_id, '_thumbnail_id', $attachment_id );
-						}
-					}
-				} else {
-					delete_post_meta( $variation_id, '_thumbnail_id' );
+			if ( isset( $data['image'] ) && is_array( $data['image'] ) ) {
+				$image = current( $data['image'] );
+				if ( is_array( $image ) ) {
+					$image['position'] = 0;
 				}
+
+				$variation = $this->save_product_images( $variation, array( $image ) );
 			}
 
 			// Virtual variation.
-			if ( isset( $variation['virtual'] ) ) {
-				$is_virtual = ( true === $variation['virtual'] ) ? 'yes' : 'no';
-				update_post_meta( $variation_id, '_virtual', $is_virtual );
+			if ( isset( $data['virtual'] ) ) {
+				$variation->set_virtual( $data['virtual'] );
 			}
 
 			// Downloadable variation.
-			if ( isset( $variation['downloadable'] ) ) {
-				$is_downloadable = ( true === $variation['downloadable'] ) ? 'yes' : 'no';
-				update_post_meta( $variation_id, '_downloadable', $is_downloadable );
+			if ( isset( $data['downloadable'] ) ) {
+				$is_downloadable = $data['downloadable'];
+				$variation->set_downloadable( $is_downloadable );
 			} else {
-				$is_downloadable = get_post_meta( $variation_id, '_downloadable', true );
+				$is_downloadable = $variation->get_downloadable();
 			}
 
-			// Shipping data.
-			$variation_object = $this->save_product_shipping_data( $variation_object, $variation );
-
-			// Stock handling.
-			if ( isset( $variation['manage_stock'] ) ) {
-				$manage_stock = ( true === $variation['manage_stock'] ) ? 'yes' : 'no';
-			} else {
-				$manage_stock = get_post_meta( $variation_id, '_manage_stock', true );
-			}
+			// Downloads.
+			if ( $is_downloadable ) {
+				// Downloadable files.
+				if ( isset( $data['downloads'] ) && is_array( $data['downloads'] ) ) {
+					$variation = $this->save_downloadable_files( $variation, $data['downloads'], $variation->get_id() );
```

Comment:
Just a note, downloads was refactored in a recent PR. Please check this method for compatibility.